### PR TITLE
net: lib: nrf_cloud: add fota validation function 

### PR DESCRIPTION
**Pull request by jayteemo**
Add function to validate a FOTA job prior to initializing
the nrf_cloud library. This allows applications to control
the reboot/modem lib reinit process.

For atv2, properly validate an nrf cloud modem FOTA job after checking
modem lib init status.
CIA-656

Also including a trivial FOTA related commit for REST error code processing.
IRIS-4451

### CODE REVIEW

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_common.c [hunk 1]
```diff
@@ -16,6 +16,89 @@
...
+	}
+#endif
+
+	return validate;
+}
+
+enum nrf_cloud_fota_validate_status modem_fota_validate_get(void)
+{
+#if defined(CONFIG_NRF_MODEM_LIB)
+	int modem_lib_init_result = nrf_modem_lib_get_init_ret();
```
Comment (lemrey) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_common.c:79]: fyi, I am planning to deprecate this function in favor of receiving a callback upon libmodem's initialization via the `NRF_MODEM_LIB_ON_INIT` macro.

Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_common.c:79]: when do you expect that change to happen?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c [hunk 2]
```diff
@@ -82,6 +82,8 @@ enum subscription_topic_index {
 	SUB_TOPIC_IDX__COUNT,
 };
 
+enum nrf_cloud_fota_validate_status modem_fota_validate_get(const int modem_lib_init_result);
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:85]: I think this can be removed, you don't reference it in this file

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c [hunk 3]
```diff
@@ -196,38 +257,24 @@ int nrf_cloud_fota_init(nrf_cloud_fota_callback_t cb)
...
-	if (ret == 0) {
-
-		/* save status and update when cloud connection is ready */
-		save_validate_status(saved_job.id, saved_job.type, saved_job.validate);
-
-		if (reboot_required) {
-			LOG_INF("Rebooting to complete FOTA update...");
-			sys_reboot(SYS_REBOOT_COLD);
-		}
+	ret = nrf_cloud_fota_pending_job_validate(NULL);
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:260]: Due to the change in the meaning of the return value in `nrf_cloud_fota_pending_job_validate` and in `nrf_cloud_fota_init`, setting `ret` directly like this makes the logic hard to follow, even though it is technically less verbose

Perhaps declare a separate variable to receive this value from `nrf_cloud_fota_pending_job_validate` and set `ret` accordingly?

**Review by glarsennordic — APPROVED** [review]
LGTM, other than a suggestion about code clarity, and what appears to be a an unneeded forward declaration

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c (hunk 3 above)
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:258]: added comment to explain `ret`

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c [hunk 4]
```diff
@@ -171,10 +173,67 @@ static int fota_settings_set(const char *key, size_t len_rd,
...
+	}
+
+	/* Do not enforce a reboot for modem updates since they can also be handled by
+	 * reinitializing the modem lib
+	 */
+	if (saved_job.type != NRF_CLOUD_FOTA_MODEM) {
+		reboot_on_init = reboot;
+	}
+
+	if (save_validate_status(saved_job.id, saved_job.type, saved_job.validate)) {
```
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota.c:217]: if this function fails i don't think there is much that can be done.
the job status will still be updated on nrf cloud, it will just be noted as "not validated".

File: doc/nrf/releases/release-notes-changelog.rst [hunk 5]
```diff
@@ -267,8 +271,14 @@ Libraries for networking
   * :ref:`lib_nrf_cloud` library:
 
     * Fixed:
+
       * An issue that caused the application to receive multiple disconnect events.
 
+    * Added:
+
+      * :c:func:`nrf_cloud_fota_pending_job_validate` function that enables an application to validate a pending FOTA job prior to initializing the :ref:`lib_nrf_cloud` library.
+      * Handling for new nRF Cloud REST error code 40499. Moved error log from :c:func:`nrf_cloud_parse_rest_error` into the calling function.
```
Comment (peknis) [inline doc/nrf/releases/release-notes-changelog.rst:280]:
```suggestion
      * :c:func:`nrf_cloud_fota_pending_job_validate` function that enables an application to validate a pending FOTA job before initializing the :ref:`lib_nrf_cloud` library.
      * Handling for new nRF Cloud REST error code 40499.
        Moved the error log from the :c:func:`nrf_cloud_parse_rest_error` function into the calling function.
```

File: include/net/nrf_cloud.h [hunk 6]
```diff
@@ -689,6 +691,25 @@ int nrf_cloud_handle_error_message(const char *const buf,
 				   const char *const msg_type,
 				   enum nrf_cloud_error *const err);
 
+/**
+ * @brief Function to validate a pending FOTA installation prior to initializing this library.
```
Comment (peknis) [inline include/net/nrf_cloud.h:695]:
```suggestion
 * @brief Function to validate a pending FOTA installation before initializing this library.
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Review by peknis — APPROVED** [review]

**Review by simensrostad — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@jayteemo Can you resolve the conflict?

**Comment by jayteemo** [comment]
> @jayteemo Can you resolve the conflict?

Conflict resolved.
